contrib/src/main/java/gov/nasa/jpl/aerie/contrib/streamline/modeling/polynomial/LinearBoundaryConsistencySolver.java

```diff
@@ -327,9 +323,50 @@ private Stream<DirectionalConstraint> directionalConstraints(Variable constraine
       }, drivingVariables));
     }
   }
+
   // Directional constraints are useful for arc consistency, since they have input (driving) and output (constrained) variables.
   // However, many directional constraints are required in general to express one General constraint.
-  private record DirectionalConstraint(Variable constrainedVariable, InequalityComparison comparison, Function<Map<Variable, ? extends Domain>, Expiring<Polynomial>> bound, Set<Variable> drivingVariables) {}
+  private static final class DirectionalConstraint {
+    private final Variable constrainedVariable;
+    private final InequalityComparison comparison;
+    private final Function<Map<Variable, ? extends Domain>, Expiring<Polynomial>> bound;
+    private final Set<Variable> drivingVariables;
+
+    private DirectionalConstraint(Variable constrainedVariable, InequalityComparison comparison, Function<Map<Variable, ? extends Domain>, Expiring<Polynomial>> bound, Set<Variable> drivingVariables) {
+      this.constrainedVariable = constrainedVariable;
+      this.comparison = comparison;
+      this.bound = bound;
+      this.drivingVariables = drivingVariables;
+    }
+
+    public Variable constrainedVariable() {
+      return constrainedVariable;
+    }
+
+    public InequalityComparison comparison() {
+      return comparison;
+    }
+
+    public Function<Map<Variable, ? extends Domain>, Expiring<Polynomial>> bound() {
+      return bound;
+    }
+
+    public Set<Variable> drivingVariables() {
+      return drivingVariables;
+    }
+
+    // *Don't* override hashCode and equals - we want to use object identity here instead of value equality.
+    // This makes it more performant during solving to check whether a constraint is already in the queue to be propagated.
```

Comment:
I appreciate the comment. It does make me wonder whether we're missing out on the ability to prune distinct, but equivalent constraints. Perhaps that's rare enough not to warrant special consideration.

Reply:
Just me thinking-in-text. TL;DR - I think there's something here, but I'm not sure it's worth chasing until we see a real use case that's bottlenecked by the solver.

My first guess would be that that's rare... The constraint expansion basically re-arranges the constraints given by the user to phrase them in terms of an expression constraining each involved variable. If you give a set of independent constraints, I think the expansions should all be independent too, at least to start. It's possible that one constraint will dominate another over some restriction of the domains, though. Consider "X <= Y + 10" and "X <= 2Y". For Y <= 5, the latter controls X, for Y >= 5 the former does. If at some point you restricted the domain for Y to just one of those conditions, say you find that Y is in [7, 20], then at least in theory you could throw away the non-dominant constraint for the rest of this solving iteration. The fact that the constraints are linear ought to help here. There's just one or two levels too many of linear algebra for me to fully complete that thought at the moment, though...

Kind of expanding on that idea though... It may also be possible to determine something about when a constraint is "tight" (i.e., re-evaluating it might actually restrict a domain further) vs. "slack" (it's not the limiting constraint, no need to re-evaluate), even if neither one is strictly dominant like in the prior example. I don't know if the bookkeeping to maintain that would be more or less expensive than just running the restriction process, though. I kind of suspect that level of granularity is equivalent to just propagating the constraint.

The other common way of optimizing a solver like this would be constraint ordering. If we could sort the constraints well, we might be able to reduce the number of constraint propagations we do. Something like "the constraint with the most variable domains that have changed is propagated next", or "the constraint with the most restricted domains is propagated next" might be good enough. Note that we don't have to do backtracking, so the usual search heuristics kind of don't apply.

Reply:
As I was thinking through this a bit more, I think there's one easier optimization we could make without too much effort. Since all constraints are linear inequalities, they do the restriction by looking at one endpoint of each variables' domain, either high or low. Currently, we add constraints to re-propagate whenever an involved domain changes, but we don't take into account which endpoint of the domain changed. I.e., "X <= 2Y" only cares about the high end of Y's domain, but currently, if Y were restricted from [0, 10] to [5, 10], we'd still queue this constraint to be re-propagated.

If the keys in `neighboringConstraints` were not just variables, but variable + which side of the domain changed, we could probably throw out at least some irrelevant constraints with minimal extra effort.